bliss/simulator/simulated_dataset.py

```diff
@@ -87,3 +91,76 @@ def val_dataloader(self):
 
     def test_dataloader(self):
         return DataLoader(self, batch_size=None, num_workers=self.num_workers)
+
+
+class CachedSimulatedDataset(pl.LightningDataModule, IterableDataset):
+    def __init__(
+        self,
+        n_batches: int,
+        valid_n_batches: int,
+        fix_validation_set: bool,
+        num_workers: int,
+        batch_size: int,
+        file_data_capacity: int,
+        cached_data_path: str,
+    ):
+        super().__init__()
+
+        self.n_batches = n_batches
+        self.valid_n_batches = valid_n_batches
+        self.fix_validation_set = fix_validation_set
+        self.num_workers = num_workers
+        self.batch_size = batch_size
+        self.file_data_capacity = file_data_capacity
+        self.cached_data_path = cached_data_path
+
+        self.data: List[FileDatum] = []
+        self.valid: Union[List[FileDatum], CachedSimulatedDataset] = []
+
+        # assume cached image files exist, read from disk
+        for filename in os.listdir(self.cached_data_path):
+            if "valid" in filename:
+                continue
+            if filename.startswith("dataset") and filename.endswith(".pkl"):
+                self.data += self.read_file(f"{self.cached_data_path}/{filename}")
+        assert self.data, "No cached data loaded; run `generate.py` first"
+
+        if self.fix_validation_set:
+            assert os.path.exists(
+                f"{self.cached_data_path}/dataset_valid.pkl"
+            ), "No cached validation data found; run `generate.py` with 'fix_validation_set=true'"
+            self.valid = self.read_file(f"{self.cached_data_path}/dataset_valid.pkl")
+        else:
+            self.valid = self
+
+    def read_file(self, filename: str) -> List[FileDatum]:
+        with open(filename, "rb") as f:
+            return pickle.load(f)
+
+    def get_batch(self, batch_idx) -> Dict:
+        batch_data = self.data[batch_idx * self.batch_size : (batch_idx + 1) * self.batch_size]
+        images = torch.stack([datum["images"] for datum in batch_data])
+        background = torch.stack([datum["background"] for datum in batch_data])
+        tile_catalog = {}
+        assert len(batch_data), f"Batch for batch_idx {batch_idx} is empty"
+        for key in batch_data[0]["tile_catalog"]:
+            tile_catalog[key] = torch.stack([datum["tile_catalog"][key] for datum in batch_data])
+        return {
+            "tile_catalog": tile_catalog,
+            "images": images,
+            "background": background,
+        }
+
+    def __iter__(self):
+        random.shuffle(self.data)
+        for batch_idx in range(self.n_batches):
+            yield self.get_batch(batch_idx)
+
+    def train_dataloader(self):
+        return DataLoader(self, batch_size=None, num_workers=0)
+
+    def val_dataloader(self):
+        return DataLoader(self.valid, batch_size=None, num_workers=0)
+
+    def test_dataloader(self):
+        return DataLoader(self, batch_size=None, num_workers=0)
```

Comment:
Is the test set the same as the training set? That seems like a problem. We should probably have separate files that contain the test set.

Reply:
Done.